### PR TITLE
fixes meteorshot throwing knocked down targets in random directions, knockback is easier to add to other projectiles codewise

### PR DESCRIPTION
**Pull request by roarbark**
## What Does This PR Do

meteorshot used to work by checking the projectiles last location before hitting and then throwing away from that. this made it throw into a random direction for horizontal targets, since the last location was ontop of them.
also shuffles the code around so if you want to give a projectiles knockback you can now more easily do it by giving it the respective variables, also a variable called "breaching" that allows you to decide if it works only on mobs or if it can throw stuff like items, chairs, tables, airlocks, windows too

## Why It's Good For The Game

meteorshot used to be able to throw a body back into you, the same way the meteorslug approached from, knocking you down. not good. now it doesnt do that anymore, always throws it away from the shooter, predictably.

also i added the variables to easily give other projectiles knockback, instead of all the code running under bullets.dm firstly because it seems more intuitive and secondly cuz i need it for my next pr

## Images of changes

BEFORE THE CHANGE

https://user-images.githubusercontent.com/41054538/198717154-5aab1093-083f-4cdc-b154-a6dfb2f618ff.mp4

AFTER THE CHANGE

https://user-images.githubusercontent.com/41054538/198718466-64c48f14-fbcd-4169-838e-b82256d2e979.mp4

## Testing

i tested the breaching var too

## Changelog
:cl:
fix: meteorshot no longer throws knocked down targets into random directions
/:cl:

### CODE REVIEW

File: config/example/config.toml [hunk 1]
```diff
@@ -145,7 +145,7 @@ ipc_screens = [
 # Enable/disable the database on a whole
 sql_enabled = false
 # SQL version. If this is a mismatch, round start will be delayed
-sql_version = 44
+sql_version = 42
 # SQL server address. Can be an IP or DNS name
 sql_address = "127.0.0.1"
```
Comment (hal9000PR) [inline config/example/config.toml:150]: revert this

**Review by hal9000PR — CHANGES_REQUESTED** [review]

File: config/example/config.toml (hunk 1 above)
Comment (roarbark) [reply config/example/config.toml]: im so sorry it wont happen again

File: code/modules/projectiles/projectile.dm [hunk 2]
```diff
@@ -117,6 +121,18 @@
...
+		if(!breaching && !isliving(target))
+			return
+		var/atom/movable/M = target
+		if(M.move_resist < INFINITY)
+			var/turf/start_turf = starting || get_turf(src)
+			var/throw_dir = get_dir(start_turf, get_turf(target))
+			var/atom/throw_target = get_edge_target_turf(get_turf(src), throw_dir)
+			M.throw_at(throw_target, knockbackdistance, knockbackspeed)
+		else
+			return
```
Comment (S34NW) [inline code/modules/projectiles/projectile.dm:134]:
```suggestion
		if(M.move_resist >= INFINITY)
		    return
		    
		var/turf/start_turf = starting || get_turf(src)
		var/throw_dir = get_dir(start_turf, get_turf(target))
		var/atom/throw_target = get_edge_target_turf(get_turf(src), throw_dir)
		M.throw_at(throw_target, knockbackdistance, knockbackspeed)			
```

Early return here please :)

Comment (S34NW) [reply code/modules/projectiles/projectile.dm:134]: Actually wait no, wtf are you returning for here. please hold

Comment (S34NW) [inline code/modules/projectiles/projectile.dm:134]:
```suggestion
		if(breaching && isliving(target))
		    var/atom/movable/M = target
		    if(M.move_resist < INFINITY)
			    var/turf/start_turf = starting || get_turf(src)
			    var/throw_dir = get_dir(start_turf, get_turf(target))
			    var/atom/throw_target = get_edge_target_turf(get_turf(src), throw_dir)
			    M.throw_at(throw_target, knockbackdistance, knockbackspeed)
```

Comment (S34NW) [reply code/modules/projectiles/projectile.dm:134]: We don't want to return here, as it will stop the rest of the hit code from running

Comment (roarbark) [reply code/modules/projectiles/projectile.dm]: the code is fixed now thanks to sean

**Review by S34NW — APPROVED** [review]

File: code/modules/projectiles/projectile.dm [hunk 3]
```diff
@@ -57,6 +57,16 @@
 	var/forcedodge = 0 //to pass through everything
 	var/dismemberment = 0 //The higher the number, the greater the bonus to dismembering. 0 will not dismember at all.
 	var/impact_effect_type //what type of impact effect to show when hitting something
+
+	///Does the projectile knock back mobs on hit?
+	var/knockback = FALSE
+	///Does the projectile throw back objects with move resist less than INFINITE on hit?
+	var/breaching = FALSE
```
Comment (farie82) [inline code/modules/projectiles/projectile.dm:64]: I feel like these two can be combined into one var using bit flags.
so something like this:
```dm
#define PROJECTILE_KNOCKBACK_MOB (1<<0)
#define PROJECTILE_KNOCKBACK_OBJ (1<<1)
...

knockback = PROJECTILE_KNOCKBACK_MOB | PROJECTILE_KNOCKBACK_OBJ
...
if(((knockback & PROJECTILE_KNOCKBACK_MOB) && isliving(target)) || ((knockback & PROJECTILE_KNOCKBACK_OBJ) && isobj(target)))
  do knockback stuff...
```

The defines would be in one of the `__DEFINE` files, or a new one if none of the existing files are suitable. This way all other files can use the defines.

File: code/modules/projectiles/projectile.dm [hunk 4]
```diff
@@ -117,6 +127,17 @@
 		return 0
 	if(alwayslog)
 		add_attack_logs(firer, target, "Shot with a [type]")
+	//If the projectile has knockback AND EITHER breaching OR the target is a mob, throw the target
+	//OR If the projectile has breaching AND the target is an object, throw the target
+	//Should the projectile hit a mob without having knockback or an object without being breaching, do not throw the target.
+	if((knockback && (breaching || isliving(target))) || (breaching && isobj(target)))
```
Comment (farie82) [inline code/modules/projectiles/projectile.dm:133]: Can you make this into a proc to make the current proc a bit more readable?
`try_apply_knockback(...)` or something among those lines will work

File: code/modules/projectiles/projectile.dm [hunk 5]
```diff
@@ -57,6 +57,16 @@
...
 	var/impact_effect_type //what type of impact effect to show when hitting something
+
+	///Does the projectile knock back mobs on hit?
+	var/knockback = FALSE
+	///Does the projectile throw back objects with move resist less than INFINITE on hit?
+	var/breaching = FALSE
+	///How far do knocked back entities travel?
+	var/knockbackdistance = 0
+	///How quickly do they make that distance?
+	var/knockbackspeed = 0
```
Comment (farie82) [inline code/modules/projectiles/projectile.dm:68]:
```suggestion
	var/knockback_distance = 0
	///How quickly do they make that distance?
	var/knockback_speed = 0
```
Be sure to use snake_case for var and proc names

**Review by farie82 — CHANGES_REQUESTED** [review]

**Comment by farie82** [comment]
Please address the open reviews within 7 days else we'll close this PR for inactivity. If this happens then you're free to reopen it at a later time.

**Comment by hal9000PR** [comment]
Feel free to reopen with requested changes

**Pull request closed without merging**